planner/core/find_best_task.go

```diff
@@ -196,6 +197,136 @@ func (ds *DataSource) tryToGetDualTask() (task, error) {
+}
+
+func compareBool(l, r bool) int {
+	if l == r {
+		return 0
+	}
+	if l == false {
+		return -1
+	}
+	return 1
+}
+
+// compareCandidates is the core of skyline pruning. It compares the two candidate paths on three dimensions:
+// the set of columns that occurred in the access condition, whether or not it matches the physical property
+// and does it require a double scan.  If `x` is not worse than `y` at all factors,
+// and there exists one factor that `x` is better than `y`, then `x` is better than `y`.
+func compareCandidates(lhs, rhs *candidatePath) int {
+	setsResult, comparable := compareColumnSet(lhs.columnSet, rhs.columnSet)
+	if !comparable {
+		return 0
+	}
+	scanResult := compareBool(lhs.singleScan, rhs.singleScan)
+	matchResult := compareBool(lhs.matchProp, rhs.matchProp)
+	sum := setsResult + scanResult + matchResult
+	if setsResult >= 0 && scanResult >= 0 && matchResult >= 0 && sum > 0 {
+		return 1
+	}
+	if setsResult <= 0 && scanResult <= 0 && matchResult <= 0 && sum < 0 {
+		return -1
+	}
+	return 0
+}
+
+func (ds *DataSource) getTableCandidate(path *accessPath, prop *property.PhysicalProperty) *candidatePath {
+	point := &candidatePath{path: path}
+	if ds.tableInfo.PKIsHandle {
+		if pkColInfo := ds.tableInfo.GetPkColInfo(); pkColInfo != nil {
+			pkCol := expression.ColInfo2Col(ds.schema.Columns, pkColInfo)
+			point.matchProp = len(prop.Items) == 1 && pkCol != nil && prop.Items[0].Col.Equal(nil, pkCol)
+		}
+	}
+	point.columnSet = expression.ExtractColumnSet(path.accessConds)
+	point.singleScan = true
+	return point
+}
+
+func (ds *DataSource) getIndexCandidate(path *accessPath, prop *property.PhysicalProperty) *candidatePath {
+	point := &candidatePath{path: path}
+	all, _ := prop.AllSameOrder()
+	if !prop.IsEmpty() && all {
+		for i, col := range path.index.Columns {
+			// not matched
+			if col.Name.L == prop.Items[0].Col.ColName.L {
+				point.matchProp = matchIndicesProp(path.index.Columns[i:], prop.Items)
+				break
+			} else if i >= path.eqCondCount {
+				break
+			}
+		}
+	}
+	point.columnSet = expression.ExtractColumnSet(path.accessConds)
+	point.singleScan = isCoveringIndex(ds.schema.Columns, path.index.Columns, ds.tableInfo.PKIsHandle)
+	return point
+}
+
+func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candidatePath {
+	candidates := make([]*candidatePath, 0, 4)
+	for _, path := range ds.possibleAccessPaths {
+		// if we already know the range of the scan is empty, just return a TableDual
+		if len(path.ranges) == 0 && !ds.ctx.GetSessionVars().StmtCtx.UseCache {
+			return []*candidatePath{{path: path}}
+		}
+		var currentCandidate *candidatePath
+		if path.isTablePath {
+			currentCandidate = ds.getTableCandidate(path, prop)
+		} else {
+			// We will use index to generate physical plan if:
+			// this path's access cond is not nil or
+			// we have prop to match or
+			// this index is forced to choose.
+			if len(path.accessConds) > 0 || len(prop.Items) > 0 || path.forced {
```

Comment:
This `if` can be put into line 300?

Reply:
Line 300 will make the comment appears in the `table scan` block.

Reply:
Keep this comment in `else if` block is acceptable.